beta/src/content/apis/react/memo.md

```diff
@@ -2,21 +2,314 @@
+  return <button>{title}</button>
+});
+```
 
+</Intro>
 
-<Intro>
+- [Usage](#usage)
+  - [Avoiding re-rendering when props are unchanged](#avoiding-rerendering-when-props-are-unchanged)
+  - [Updating a memoized component using state](#updating-a-memoized-component-using-state)
+  - [Updating a memoized component using a context](#updating-a-memoized-component-using-state)
+  - [Minimizing props changes](#minimizng-props-changes)
+  - [Specifying a custom comparison function](#specifying-a-custom-comparison-function)
+- [Reference](#reference)
+- [Troubleshooting](#troubleshooting)
+  - [My component re-renders when a prop is an object or array](#my-component-rerenders-when-a-prop-is-an-object-or-array)
+  - [My component re-renders when a prop is a function](#my-component-rerenders-when-a-prop-is-a-function)
+
+---
+
+## Usage {/*usage*/}
+
+### Avoiding re-rendering when props are unchanged {/*avoiding-rerendering-when-props-are-unchanged*/}
+
+React normally re-renders a component whenever its parent re-renders. With `memo`, you can create a component that React will not re-render when its parent re-renders so long as its new props are the same as the old props. Such a component is said to be *memoized*.
+
+`memo` is a function: you use it by wrapping your component in a call to `memo` and using the value that it returns in place of your original component:
+
+```
+export default memo(function Greeting({name}) {
+  return <>Hello, {name}!</>;
+});
+```
+
+A React component should always render the same output if its props, state, and context haven't changed. By using `memo`, you are telling React that your component complies with this requirement, so React doesn't need to re-render as long as none of those things have changed. When you use `memo`, your component will still re-render if its own state changes or if a context that it's using changes.
+
+<Gotcha>
+
+Use `memo` only as a performance optimization, not to guarantee that a re-render
+will not happen. React may choose to re-render anyway in some situations, such as to save
+memory when a component is off-screen.
+
+</Gotcha>
+
+
+In this example, notice that the `Greeting` component re-renders whenever `name` is changed (because that's one of its props), but not when `address` is changed (because it's not a prop):
+
+<Sandpack>
+
+```js
+import { memo, useEffect, useState } from 'react';
+
+export default function MyApp() {
+  const [name, setName] = useState('');
+  const [address, setAddress] = useState('');
+  return (
+    <>
+      <label>
+        Name{': '}
+        <input value={name} onChange={e => setName(e.target.value)} />
+      </label>
+      <label>
+        Address{': '}
+        <input value={address} onChange={e => setAddress(e.target.value)} />
+      </label>
+      <Greeting name={name} />
+    </>
+  );
+}
+
+const Greeting = memo(function Greeting({name}) {
+  useEffect(() => {
+    console.log("Greeting was rendered");
```

Comment:
This is a bit misleading since it really shows the timing of effects. I think it would make sense to place the log in the body directly. It might be good to do `rendered at [...]` with a timestamp or something so that it's easier to distinguish repeated renders

Reply:
I'm not too happy about this because I'm concerned that we're showing an example of a side-effect in render. Also, either way you get the initial logs on page load plus there are lots of renders (4) per change, which I could see could confuse the issue.

Can you think of a better way to indicate when the render is happening or not happening?

Reply:
I think for practical reasons we don't consider console logs a side effect -- it's a debugging mechanism and we do expect people to log during render